File: lib/cfr/builtins/libs/flutter/rendering/containerBoxParentData.dart
import 'dart:core';
import 'dart:ui';

import 'package:flutter/src/rendering/box.dart';
import 'package:flutter/src/rendering/object.dart';

import 'package:hydro_sdk/cfr/runtimeSupport.dart';

class VMManagedContainerBoxParentData
    extends VMManagedBox<ContainerBoxParentData<RenderObject>> {
  VMManagedContainerBoxParentData(
      {required this.table, required this.vmObject, required this.hydroState})
      : super(
          table: table,
          vmObject: vmObject,
          hydroState: hydroState,
        ) {
    table['offset'] = maybeBoxObject<Offset>(
        object: vmObject.offset, hydroState: hydroState, table: HydroTable());
    table['detach'] = makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      vmObject.detach();
      return [];
    });
    table['toString'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.toString(),
      ];
    });
    table['getHashCode'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.hashCode,
      ];
    });
  }

  final HydroTable table;

  final HydroState hydroState;

  final ContainerBoxParentData<RenderObject> vmObject;
}

class RTManagedContainerBoxParentData extends ContainerBoxParentData
    implements Box<ContainerBoxParentData> {
  RTManagedContainerBoxParentData(
      {required this.table, required this.hydroState})
      : super() {
    table['vmObject'] = vmObject;
    table['unwrap'] = makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [unwrap()];
    });

    table['offset'] = maybeBoxObject<Offset>(
        object: offset, hydroState: hydroState, table: HydroTable());
    table['_dart_detach'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      super.detach();
      return [];
    });
    table['_dart_toString'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [super.toString()];
    });
    table['_dart_getHashCode'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [super.hashCode];
    });
  }

  final HydroTable table;

  final HydroState hydroState;

  ContainerBoxParentData unwrap() => this;
  ContainerBoxParentData get vmObject => this;
  @override
  void detach() {
    super.detach();
    Closure closure = table["detach"];
    return closure.dispatch([table], parentState: hydroState)[0];
  }

  @override
  String toString() {
    Closure closure = table["__tostring"];
    return closure.dispatch([table], parentState: hydroState)[0];
  }

  @override
  int get hashCode {
    Closure closure = table["getHashCode"];
    return closure.dispatch([table], parentState: hydroState)[0];
  }
}

void loadContainerBoxParentData(
    {required HydroState hydroState, required HydroTable table}) {
  table['containerBoxParentData'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      RTManagedContainerBoxParentData(
          table: luaCallerArguments[0], hydroState: hydroState)
    ];
  });
  registerBoxer<ContainerBoxParentData>(boxer: (
      {required ContainerBoxParentData vmObject,
      required HydroState hydroState,
      required HydroTable table}) {
    return VMManagedContainerBoxParentData(
        vmObject: vmObject, hydroState: hydroState, table: table);
  });
}

File: runtime/flutter/rendering/containerBoxParentData.ts
import { IOffset } from "../../dart/ui/offset";
import { IBoxParentData } from "./boxParentData";
import { IContainerParentDataMixin } from "./containerParentDataMixin";
import { IRenderObject } from "./renderObject";
declare const flutter: {
    rendering: {
        containerBoxParentData: <ChildType>(
            this: void,
            containerBoxParentData: IContainerBoxParentData<ChildType>
        ) => IContainerBoxParentData<ChildType>;
    };
};
export interface IContainerBoxParentData<ChildType> {
    previousSibling: ChildType | undefined;
    nextSibling: ChildType | undefined;
    offset: IOffset;
    detach: () => void;
    toString: () => string;
    getHashCode: () => number;
}
export class ContainerBoxParentData<ChildType>
    implements IBoxParentData, IContainerParentDataMixin<ChildType>
{
    public readonly previousSibling: ChildType | undefined = undefined as any;
    public readonly nextSibling: ChildType | undefined = undefined as any;
    public readonly offset: IOffset = undefined as any;
    public constructor() {
        flutter.rendering.containerBoxParentData(this);
    }
    private readonly _dart_detach: () => void = undefined as any;
    private readonly _dart_toString: () => string = undefined as any;
    private readonly _dart_getHashCode: () => number = undefined as any;
    public detach(): void {
        return this._dart_detach();
    }
    public toString(): string {
        return this._dart_toString();
    }
    public getHashCode(): number {
        return this._dart_getHashCode();
    }
}
